packages/vue/src/utils.ts

```diff
@@ -89,3 +90,109 @@ export function transformComposableProps<T extends { context: object }>(props: T
     context: reactive(props.context),
   }
 }
+
+/**
+ * Checks whether a given VNode is a render-vialble element.
+ */
+export function isValidVNodeElement(input: any): boolean {
+  // No children
+  if (input == null) return false
+  // Raw HTML elements like 'div', 'span', e.t.c.
+  if (typeof input.type === 'string') return true
+  // Other Vue component objects
+  if (typeof input.type === 'object') return true
+  // Functional components
+  if (typeof input.type === 'function') return true
+  // HTML Comment nodes or Text nodes
+  return false
+}
+
+/**
+ * Recursively flattens the Fragment descendants of a given VNode.
+ *
+ * When you create a component and pass a <slot />, Vue wraps
+ * the contents of <slot /> inside a <Fragment /> component and assigns
+ * the <slot /> VNode a type of Fragment.
+ *
+ * So why are we flattening here? Vue renders VNodes from the leaf
+ * nodes going up to the root. In other words, when executing the render function
+ * of each component, it executes the child render functions first before the parents.
+ *
+ * This means that at any components render function execution context, all it's children
+ * VNodes should have already been rendered -- and that includes any slots! :D
+ *
+ * In the cases where we pass in a component with slots to the `asChild` component,
+ * we shall need to flatten those slot fragment VNodes so as to extract all it's children VNodes
+ * to correctly apply the props and event listeners from the with as child components.
+ *
+ * We do this recursively to ensure that all first child slots that contain fragments in their descendants are rendered into VNodes before passing events.
+ * to the first actual element VNode.
+ */
+export function renderSlotFragments(children: VNode[]): VNode[] {
+  return children.flatMap((child) => {
+    if (child.type === Fragment) {
+      return renderSlotFragments(child.children as VNode[])
+    }
+    return [child]
+  })
+}
+
+/**
+ * We use our own `mergeProps` instead of Vue's mergeProps to correctly handle event listeners.
+ *
+ * Credit: headless-ui/vue: https://github.com/tailwindlabs/headlessui/blob/main/packages/@headlessui-vue/src/utils/render.ts#L196
+ */
+export function mergeProps(...propsArgs: Record<any, any>[]) {
```

Comment:
We also want to merge styles and classnames applied to the elements. Please see the merge-props we use in React: https://github.com/chakra-ui/zag/blob/main/packages/core/src/merge-props.ts